File: scripts/translate_dramaclaw_pt.py
"""Generate the pt-BR DramaClaw locale using the already configured Gemini provider."""

import json
import sys
import tomllib
from concurrent.futures import ThreadPoolExecutor, as_completed
from pathlib import Path

from google import genai
from google.genai import types

ROOT = Path(__file__).resolve().parents[1]
SOURCE = ROOT / "engines/dramaclaw/frontend/public/locales/en/translation.json"
TARGET = ROOT / "engines/dramaclaw/frontend/public/locales/pt/translation.json"
CONFIG = ROOT / "engines/moneyprinter/config.toml"

source = json.loads(SOURCE.read_text())
app_config = tomllib.loads(CONFIG.read_text()).get("app", {})
api_key = app_config.get("gemini_api_key", "")
model = app_config.get("gemini_model_name", "gemini-2.5-flash")
if not api_key:
    raise SystemExit("Gemini não está configurado no MoneyPrinterTurbo")

strings: list[str] = []


def collect(value):
    if isinstance(value, str):
        strings.append(value)
    elif isinstance(value, list):
        for child in value:
            collect(child)
    elif isinstance(value, dict):
        for child in value.values():
            collect(child)


collect(source)
unique = list(dict.fromkeys(strings))
chunks = [unique[index:index + 220] for index in range(0, len(unique), 220)]


def translate_chunk(index_and_chunk):
    index, chunk = index_and_chunk
    payload = {str(position): value for position, value in enumerate(chunk)}
    prompt = (
        "Traduza todos os valores do JSON para português brasileiro natural, "
        "premium e conciso para um software profissional de produção audiovisual. "
        "Preserve exatamente placeholders {{...}}, tags, URLs, atalhos, nomes de "
        "modelos, extensões e códigos. Não altere as chaves numéricas. Retorne "
        "somente o objeto JSON completo.\n" + json.dumps(payload, ensure_ascii=False)
    )
    for attempt in range(3):
        try:
            with genai.Client(api_key=api_key) as client:
                response = client.models.generate_content(
                    model=model,
                    contents=prompt,
                    config=types.GenerateContentConfig(
                        temperature=0,
                        response_mime_type="application/json",
                        max_output_tokens=16384,
                    ),
                )
            output = json.loads(response.text)
            if len(output) != len(chunk):
                raise ValueError("quantidade de textos diferente do lote")
            print(f"lote {index + 1}/{len(chunks)} traduzido", flush=True)
            return {original: str(output.get(str(position), original)) for position, original in enumerate(chunk)}
        except Exception as exc:
            if attempt == 2:
                print(f"lote {index + 1} com fallback: {type(exc).__name__}", file=sys.stderr, flush=True)
                return {original: original for original in chunk}


translated: dict[str, str] = {}
with ThreadPoolExecutor(max_workers=3) as executor:
    futures = [executor.submit(translate_chunk, item) for item in enumerate(chunks)]
    for future in as_completed(futures):
        translated.update(future.result())


def rebuild(value):
    if isinstance(value, str):
        return translated.get(value, value)
    if isinstance(value, list):
        return [rebuild(child) for child in value]
    if isinstance(value, dict):
        return {key: rebuild(child) for key, child in value.items()}
    return value


TARGET.parent.mkdir(parents=True, exist_ok=True)
TARGET.write_text(json.dumps(rebuild(source), ensure_ascii=False, indent=2) + "\n")
print(f"pt-BR concluído: {len(unique)} textos", flush=True)

File: scripts/translate-dramaclaw-pt.mjs
import fs from 'node:fs/promises';

const sourcePath = new URL('../engines/dramaclaw/frontend/public/locales/en/translation.json', import.meta.url);
const targetPath = new URL('../engines/dramaclaw/frontend/public/locales/pt/translation.json', import.meta.url);
const checkpointPath = new URL('../.runtime/dramaclaw-pt-checkpoint.json', import.meta.url);
const source = JSON.parse(await fs.readFile(sourcePath, 'utf8'));

const strings = [];
function collect(value) {
  if (typeof value === 'string') strings.push(value);
  else if (Array.isArray(value)) value.forEach(collect);
  else if (value && typeof value === 'object') Object.values(value).forEach(collect);
}
collect(source);
const unique = [...new Set(strings)];
const translated = new Map();
try {
  const checkpoint = JSON.parse(await fs.readFile(checkpointPath, 'utf8'));
  Object.entries(checkpoint).forEach(([original, value]) => translated.set(original, String(value)));
} catch {}

const chunks = [];
const pending = unique.filter((value) => !translated.has(value));
for (let index = 0; index < pending.length; index += 140) chunks.push(pending.slice(index, index + 140));
let saveChain = Promise.resolve();

function saveCheckpoint() {
  saveChain = saveChain.then(() => fs.writeFile(checkpointPath, JSON.stringify(Object.fromEntries(translated))));
  return saveChain;
}

async function translateChunk(chunk, chunkIndex) {
  const input = Object.fromEntries(chunk.map((value, index) => [String(index), value]));
  const prompt = `Traduza os valores deste JSON de inglês para português brasileiro natural e conciso para uma interface profissional de produção audiovisual. Preserve exatamente placeholders como {{name}}, tags, atalhos, URLs, nomes de modelos, extensões e códigos. Não altere as chaves numéricas. Responda somente com um objeto JSON válido com todas as chaves.\n${JSON.stringify(input)}`;
  for (let attempt = 1; attempt <= 3; attempt += 1) {
    try {
      const response = await fetch('http://127.0.0.1:11434/api/generate', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ model: 'gemma3:4b', prompt, stream: false, format: 'json', options: { temperature: 0, num_ctx: 16384 } }),
      });
      if (!response.ok) throw new Error(`HTTP ${response.status}`);
      const body = await response.json();
      const output = JSON.parse(body.response);
      if (Object.keys(output).length !== chunk.length) throw new Error(`esperado ${chunk.length}, recebido ${Object.keys(output).length}`);
      chunk.forEach((original, index) => translated.set(original, String(output[String(index)] ?? original)));
      await saveCheckpoint();
      process.stdout.write(`lote ${chunkIndex + 1}/${chunks.length} traduzido\n`);
      return;
    } catch (error) {
      if (attempt === 3) {
        process.stderr.write(`lote ${chunkIndex + 1} manteve fallback em inglês: ${error}\n`);
        chunk.forEach((original) => translated.set(original, original));
        await saveCheckpoint();
        return;
      }
    }
  }
}

let next = 0;
async function worker() {
  while (next < chunks.length) {
    const index = next++;
    await translateChunk(chunks[index], index);
  }
}
await Promise.all([worker(), worker(), worker(), worker()]);

function rebuild(value) {
  if (typeof value === 'string') return translated.get(value) ?? value;
  if (Array.isArray(value)) return value.map(rebuild);
  if (value && typeof value === 'object') return Object.fromEntries(Object.entries(value).map(([key, child]) => [key, rebuild(child)]));
  return value;
}

await fs.mkdir(new URL('../engines/dramaclaw/frontend/public/locales/pt/', import.meta.url), { recursive: true });
await fs.writeFile(targetPath, `${JSON.stringify(rebuild(source), null, 2)}\n`);
process.stdout.write(`pt-BR concluído: ${unique.length} textos, ${chunks.length} lotes\n`);
